julia-mode: Move tests to new file and improve

justbur opened the pull request:
Create file julia-mode-tests.el to hold all ert tests, so that tests are
not loaded with julia-mode.el

Instead of ignoring-warnings for everyone to get around some users not
having ert, use require with the NOERROR flag.

Have julia--run-tests print a helpful message for thsoe who do not have
ert installed.

Fixes #13237

justbur committed:
julia-mode: Move tests to new file and improve

Create file julia-mode-tests.el to hold all ert tests, so that tests are
not loaded with julia-mode.el

Instead of ignoring-warnings for everyone to get around some users not
having ert, use require with the NOERROR flag.

Have julia--run-tests print a helpful message for thsoe who do not have
ert installed.

Fixes #13237

yuyichao commented:
How should the test file be installed and run after this?

c.c.  @Wilfred

justbur commented:
You can run them after loading the test file. I didn't see any scripts to
automate testing of julia-mode in the repo. How is it handled?

The point of this is to not load the tests for everyone who loads
julia-mode.el.
On Fri, Feb 19, 2016 at 6:10 PM Yichao Yu <notifications@github.com> wrote:

> How should the test file be installed and run after this?
>
> c.c. @Wilfred <https://github.com/Wilfred>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/15140#issuecomment-186444522>.
>

yuyichao commented:
No the tests are not run on any CI AFAIK. I usually copy it locally and run the test. I was wondering if there are automatic hooks to tell emacs where to find the test script but loading the file manually is also fine.

justbur commented:
With ert you can batch run the tests from the command line too.

I just realized there's a (require 'julia-mode) from the tests file. I'll
fix that later.
On Fri, Feb 19, 2016 at 6:21 PM Yichao Yu <notifications@github.com> wrote:

> No the tests are not run on any CI AFAIK. I usually copy it locally and
> run the test. I was wondering if there are automatic hooks to tell emacs
> where to find the test script but loading the file manually is also fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaLang/julia/pull/15140#issuecomment-186448332>.
>

justbur commented:
I rebased onto master and added the require along with a usage note for the command that you would need to run from the command line. Loading the file manually and running the tests is fine too.

Wilfred commented on contrib/julia-mode-tests.el:
```diff
@@ -0,0 +1,224 @@
+(require 'julia-mode)
+
+;; Emacs 23.X doesn't include ert, so we only test if ert is available.
+(when (require 'ert nil t)
```
The tests don't work on Emacs 23.X. Is it worth wrapping the entire file in this `when`? Now it's a separate test file, I'd be tempted to just do `(require 'ert)`

justbur replied:
That's a good point. I don't think it needs to be wrapped anymore. 

I'll fix it now.

justbur replied:
Removed the when

Wilfred commented:
Looks good to me!

@yuyichao running the tests should just be a matter of opening the file, `M-x eval-buffer` then `M-x julia--run-tests` (or `M-x ert`).

yuyichao commented:
LGTM and test passed locally (mainly to double check the parenthesis ;-p )
